File: AutoFiller.py
from datetime import datetime

from docx import Document
import re


class AutoFiller:
    def __init__(self, docx_file):
        self.doc = Document(docx_file)
        self.dict_values = {}

    def date_to_dict(self, tag: str, date_str: str):
        date = datetime.strptime(date_str, '%Y-%m-%d')
        month = str(date.month)
        day = str(date.day)
        if len(month) == 1:
            month = "0" + month
        if len(day) == 1:
            day = "0" + day

        self.dict_values[f"{tag}M"] = month
        self.dict_values[f"{tag}D"] = day
        self.dict_values[f"{tag}Y"] = str(date.year)

    def add_value_to_dict(self, key: str, value: str) -> None:
        self.dict_values[key] = value

    def replace_all_in_docs(self):
        for key in self.dict_values.keys():
            print(key)
            self.replace_text_in_docx(key, self.dict_values[key])

    def replace_text_in_docx(self, old_text, new_text):
        print("\nKey: " + old_text + "\n")

        for paragraph in self.doc.paragraphs:
            if old_text in paragraph.text:
                inline = paragraph.runs
                for i in inline:
                    print(i.text, end=" ::: ")
                for i in range(len(inline)):
                    if old_text in inline[i].text:
                        text = inline[i].text.replace(old_text, new_text)
                        inline[i].text = text

        for table in self.doc.tables:
            for row in table.rows:
                for cell in row.cells:
                    for paragraph in cell.paragraphs:
                        if old_text in paragraph.text:
                            inline = paragraph.runs
                            # for i in inline:
                            #     print(i.text, end=" ::: ")
                            for i in range(len(inline)):
                                if old_text in inline[i].text:
                                    text = inline[i].text.replace(old_text, new_text)
                                    inline[i].text = text
        for section in self.doc.sections:
            header = section.footer
            for paragraph in header.paragraphs:
                print("Paragraph: " + paragraph.text)
                if old_text in paragraph.text:
                    print("Old_Text: " + old_text)
                    inline = paragraph.runs
                    for i in range(len(inline)):
                        if old_text in inline[i].text:
                            text = inline[i].text.replace(old_text, new_text)
                            inline[i].text = text

    def save_doc_to_file(self, path_new_doc):
        self.doc.save(path_new_doc)

File: FilerSpring/src/main/java/com/example/filerspring/service/AutoFiller.java
package com.example.filerspring.service;

import lombok.extern.slf4j.Slf4j;
import org.apache.poi.xwpf.model.XWPFHeaderFooterPolicy;
import org.apache.poi.xwpf.usermodel.*;
import org.springframework.beans.factory.annotation.Value;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

@Slf4j
public class AutoFiller {

    private final XWPFDocument doc;

    private final Map<String, String> dictValues;

    @Value("${mongo.file.tmp.prefix}")
    private String prefix;

    @Value("${mongo.file.tmp.suffix}")
    private String suffix;

    public AutoFiller(String docxFile, Map<String, String> dictValues) throws IOException {
        FileInputStream fis = new FileInputStream(docxFile);
        this.doc = new XWPFDocument(fis);
        this.dictValues = dictValues;
        fis.close();
    }

    public AutoFiller(FileInputStream fis) throws IOException {
        this.doc = new XWPFDocument(fis);
        this.dictValues = new HashMap<>();
        fis.close();
    }

    public AutoFiller(File file, Map<String, String> dictValues) throws IOException {
        var fis = new FileInputStream(file);
        this.doc = new XWPFDocument(fis);
        this.dictValues = dictValues;
        fis.close();
    }

    public void dateToMap(String tag, String dateStr) {
        // Convert date to specific format
        // Implement date conversion logic here
    }

    public void addValueToMap(String key, String value) {
        dictValues.put(key, value);
    }

    public void replaceAllInDocx() {
        for (String key : dictValues.keySet()) {
            System.out.println(key + " : " + dictValues.get(key));
            replaceTextInDocx(key, dictValues.get(key));
        }
    }

    private void replaceTextInDocx(String oldText, String newText) {
        for (XWPFParagraph paragraph : doc.getParagraphs()) {
            String text = paragraph.getText();
            if (text.contains(oldText)) {
                for (XWPFRun run : paragraph.getRuns()) {
                    String runText = run.getText(0);
                    if (Objects.isNull(runText)) {
                        continue;
                    }
                    if (runText.contains(oldText)) {
                        runText = runText.replace(oldText, newText);
                        run.setText(runText, 0);
                    }
                }
            }
        }

        for (XWPFTable table : doc.getTables()) {
            for (XWPFTableRow row : table.getRows()) {
                for (XWPFTableCell cell : row.getTableCells()) {
                    for (XWPFParagraph paragraph : cell.getParagraphs()) {
                        String text = paragraph.getText();
                        if (text.contains(oldText)) {
                            for (XWPFRun run : paragraph.getRuns()) {
                                String runText = run.getText(0);
                                if (runText.contains(oldText)) {
                                    runText = runText.replace(oldText, newText);
                                    run.setText(runText, 0);
                                }
                            }
                        }
                    }
                }
            }
        }

        XWPFHeaderFooterPolicy headerFooterPolicy = doc.getHeaderFooterPolicy();
        if (headerFooterPolicy != null) {
            XWPFFooter footer = headerFooterPolicy.getDefaultFooter();
            for (XWPFParagraph paragraph : footer.getParagraphs()) {
                String text = paragraph.getText();
                if (text.contains(oldText)) {
                    for (XWPFRun run : paragraph.getRuns()) {
                        String runText = run.getText(0);
                        if (runText.contains(oldText)) {
                            runText = runText.replace(oldText, newText);
                            run.setText(runText, 0);
                        }
                    }
                }
            }
        }

    }

    public void saveDocToFile(String newPath) {
        try (FileOutputStream fos = new FileOutputStream(newPath)) {
            doc.write(fos);
        } catch (IOException e) {
            log.error("Could not create file {}", newPath);
        }
    }

    public File getTempFile()  {
        File tempFile;
        try {
            tempFile = File.createTempFile(prefix, suffix);
        } catch (IOException e) {
            log.warn("Could not create temp file {}", prefix + suffix);
            throw new RuntimeException(e);
        }
        try (FileOutputStream fos = new FileOutputStream(tempFile)) {
            doc.write(fos);
        } catch (IOException e) {
            log.error("Could not create file {}", tempFile.getAbsolutePath());
        }
        return tempFile;
    }
}